lib/lib-dynamodb/src/DynamoDBDocument.ts

```diff
@@ -0,0 +1,361 @@
+import { DynamoDBDocumentClient, TranslateConfig } from "./DynamoDBDocumentClient";
+import {
+  BatchExecuteStatementCommand,
+  BatchExecuteStatementCommandInput,
+  BatchExecuteStatementCommandOutput,
+} from "./commands/BatchExecuteStatementCommand";
+import { BatchGetCommand, BatchGetCommandInput, BatchGetCommandOutput } from "./commands/BatchGetCommand";
+import { BatchWriteCommand, BatchWriteCommandInput, BatchWriteCommandOutput } from "./commands/BatchWriteCommand";
+import { DeleteCommand, DeleteCommandInput, DeleteCommandOutput } from "./commands/DeleteCommand";
+import {
+  ExecuteStatementCommand,
+  ExecuteStatementCommandInput,
+  ExecuteStatementCommandOutput,
+} from "./commands/ExecuteStatementCommand";
+import {
+  ExecuteTransactionCommand,
+  ExecuteTransactionCommandInput,
+  ExecuteTransactionCommandOutput,
+} from "./commands/ExecuteTransactionCommand";
+import { GetCommand, GetCommandInput, GetCommandOutput } from "./commands/GetCommand";
+import { PutCommand, PutCommandInput, PutCommandOutput } from "./commands/PutCommand";
+import { QueryCommand, QueryCommandInput, QueryCommandOutput } from "./commands/QueryCommand";
+import { ScanCommand, ScanCommandInput, ScanCommandOutput } from "./commands/ScanCommand";
+import { TransactGetCommand, TransactGetCommandInput, TransactGetCommandOutput } from "./commands/TransactGetCommand";
+import {
+  TransactWriteCommand,
+  TransactWriteCommandInput,
+  TransactWriteCommandOutput,
+} from "./commands/TransactWriteCommand";
+import { UpdateCommand, UpdateCommandInput, UpdateCommandOutput } from "./commands/UpdateCommand";
+import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
+import { HttpHandlerOptions as __HttpHandlerOptions } from "@aws-sdk/types";
+
+export class DynamoDBDocument extends DynamoDBDocumentClient {
+  static from(client: DynamoDBClient, translateConfig?: TranslateConfig) {
+    return new DynamoDBDocument(client, translateConfig);
+  }
+
+  public batchExecuteStatement(
+    args: BatchExecuteStatementCommandInput,
+    options?: __HttpHandlerOptions
+  ): Promise<BatchExecuteStatementCommandOutput>;
+  public batchExecuteStatement(
+    args: BatchExecuteStatementCommandInput,
+    cb: (err: any, data?: BatchExecuteStatementCommandOutput) => void
+  ): void;
+  public batchExecuteStatement(
+    args: BatchExecuteStatementCommandInput,
+    options: __HttpHandlerOptions,
+    cb: (err: any, data?: BatchExecuteStatementCommandOutput) => void
+  ): void;
+  public batchExecuteStatement(
+    args: BatchExecuteStatementCommandInput,
+    optionsOrCb?: __HttpHandlerOptions | ((err: any, data?: BatchExecuteStatementCommandOutput) => void),
+    cb?: (err: any, data?: BatchExecuteStatementCommandOutput) => void
+  ): Promise<BatchExecuteStatementCommandOutput> | void {
+    const command = new BatchExecuteStatementCommand(args);
+    if (typeof optionsOrCb === "function") {
```

Comment:
This exact `optionsOrCb` guard is written in every command method on the client. I count 13 instances of the same code. Have you considered centralizing this logic in a private class method or a utility function?

Reply:
The `optionsOrCb` guard is present to help with [overloaded functions](https://www.typescriptlang.org/docs/handbook/functions.html#overloads) for backward compatibility.

How can we centralize this logic further?
I can create a feature-request for it for follow-up.

Reply:
Yeah, the guard makes sense, but the same exact guard exists in every single command function, _as well as_ in the underlying function (`this.send`). The parameters are being checked twice on the way down at runtime.

Since `this.send` is performing these checks it seems redundant to also perform the same check in every command method. It should be possible to write the underlying function type signature in such a way that that it accepts the various overload calls; that's kind of the point of overloading, to accept multiple kinds of input.

Reply:
Created a request at https://github.com/aws/aws-sdk-js-v3/issues/2104